### ISSUE TITLE
Flutter 1.9.1+hotfix.2: Unable to precache Linux binary artifacts.

### ISSUE THREAD

**Issue by jaggernod** [flutter/flutter#40148]
## Steps to Reproduce

1. Upgrade flutter to **1.9.1+hotfix.2**
2. Run `flutter precache --linux --verbose`

## Logs

```
[  +14 ms] executing: [/home/jaggernod/work/tooling/flutter/] git log -n 1 --pretty=format:%H
[  +42 ms] Exit code 0 from: git log -n 1 --pretty=format:%H
[        ] 2d2a1ffec95cc70a3218872a2cd3f8de4933c42f
[        ] executing: [/home/jaggernod/work/tooling/flutter/] git describe --match v*.*.* --first-parent --long --tags
[  +12 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.9.1+hotfix.2-0-g2d2a1ffec
[  +13 ms] executing: [/home/jaggernod/work/tooling/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +9 ms] Exit code 128 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] fatal: HEAD does not point to a branch
[  +14 ms] executing: [/home/jaggernod/work/tooling/flutter/] git rev-parse --abbrev-ref HEAD
[  +14 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] HEAD
[  +41 ms] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[   +7 ms] executing: zip -T
/home/jaggernod/work/tooling/flutter/bin/cache/downloads/storage.googleapis.com/flutter_infra/flutter/b863200c37df4ed378042de11c4e9ff34e4e58c9/linux-x64/linux-
x64-flutter.zip
[  +15 ms] Downloading linux-x64 tools...
[   +8 ms] Downloading: https://storage.googleapis.com/flutter_infra/flutter/b863200c37df4ed378042de11c4e9ff34e4e58c9/linux-x64/linux-x64-flutter.zip
[ +457 ms] Downloading linux-x64 tools... (completed in 0.5s)
[   +1 ms] "flutter precache" took 552ms.
Download failed.
URL: https://storage.googleapis.com/flutter_infra/flutter/b863200c37df4ed378042de11c4e9ff34e4e58c9/linux-x64/linux-x64-flutter.zip
Error: 404 Not Found

#0      throwToolExit (package:flutter_tools/src/base/common.dart:28:3)
#1      _attempt (package:flutter_tools/src/base/net.dart:93:7)
<asynchronous suspension>
#2      fetchUrl (package:flutter_tools/src/base/net.dart:23:36)
<asynchronous suspension>
#3      _downloadFile (package:flutter_tools/src/cache.dart:1036:37)
<asynchronous suspension>
#4      CachedArtifact._downloadArchive.<anonymous closure> (package:flutter_tools/src/cache.dart:452:17)
<asynchronous suspension>
#5      CachedArtifact._withDownloadFile (package:flutter_tools/src/cache.dart:481:26)
<asynchronous suspension>
#6      CachedArtifact._downloadArchive (package:flutter_tools/src/cache.dart:448:12)
#7      CachedArtifact._downloadZipArchive (package:flutter_tools/src/cache.dart:468:12)
#8      EngineCachedArtifact.updateInner (package:flutter_tools/src/cache.dart:615:13)
<asynchronous suspension>
#9      CachedArtifact.update (package:flutter_tools/src/cache.dart:411:11)
<asynchronous suspension>
#10     Cache.updateAll (package:flutter_tools/src/cache.dart:324:26)
<asynchronous suspension>
#11     PrecacheCommand.runCommand (package:flutter_tools/src/commands/precache.dart:64:19)
<asynchronous suspension>
#12     FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:490:18)
<asynchronous suspension>
#13     FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:407:33)
<asynchronous suspension>
#14     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#15     _rootRun (dart:async/zone.dart:1124:13)
#16     _CustomZone.run (dart:async/zone.dart:1021:19)
#17     _runZoned (dart:async/zone.dart:1516:10)
#18     runZoned (dart:async/zone.dart:1463:12)
#19     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#20     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:397:20)
#21     CommandRunner.runCommand (package:args/command_runner.dart:197:27)
<asynchronous suspension>
#22     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:402:21)
<asynchronous suspension>
#23     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#24     _rootRun (dart:async/zone.dart:1124:13)
#25     _CustomZone.run (dart:async/zone.dart:1021:19)
#26     _runZoned (dart:async/zone.dart:1516:10)
#27     runZoned (dart:async/zone.dart:1463:12)
#28     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#29     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:356:19)
<asynchronous suspension>
#30     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
#31     new Future.sync (dart:async/future.dart:224:31)
#32     CommandRunner.run (package:args/command_runner.dart:112:14)
#33     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:242:18)
#34     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:63:22)
<asynchronous suspension>
#35     _rootRun (dart:async/zone.dart:1124:13)
#36     _CustomZone.run (dart:async/zone.dart:1021:19)
#37     _runZoned (dart:async/zone.dart:1516:10)
#38     runZoned (dart:async/zone.dart:1500:12)
#39     run.<anonymous closure> (package:flutter_tools/runner.dart:61:18)
<asynchronous suspension>
#40     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#41     _rootRun (dart:async/zone.dart:1124:13)
#42     _CustomZone.run (dart:async/zone.dart:1021:19)
#43     _runZoned (dart:async/zone.dart:1516:10)
#44     runZoned (dart:async/zone.dart:1463:12)
#45     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#46     runInContext (package:flutter_tools/src/context_runner.dart:58:24)
<asynchronous suspension>
#47     run (package:flutter_tools/runner.dart:50:10)
#48     main (package:flutter_tools/executable.dart:65:9)
<asynchronous suspension>
#49     main (file:///home/jaggernod/work/tooling/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#50     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:303:32)
#51     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

```
[✓] Flutter (Channel unknown, v1.9.1+hotfix.2, on Linux, locale en_US.UTF-8)
    • Flutter version 1.9.1+hotfix.2 at /home/jaggernod/work/tooling/flutter
    • Framework revision 2d2a1ffec9 (3 days ago), 2019-09-06 18:39:49 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/jaggernod/work/android-sdk/
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /home/jaggernod/work/ide/final/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang++ 6.0.0
    • GNU Make 4.1

[✓] Android Studio (version 3.2)
    • Android Studio at /home/jaggernod/work/ide/android-studio
    • Flutter plugin version 31.3.3
    • Dart plugin version 182.5124
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] Android Studio (version 3.5)
    • Android Studio at /home/jaggernod/work/ide/final/android-studio
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (3 available)
    • SM G965F • 24906451cf1c7ece • android-arm64  • Android 9 (API 28)
    • Linux    • Linux            • linux-x64      • Linux
    • Chrome   • chrome           • web-javascript • Google Chrome 76.0.3809.132 


• No issues found!
```

**Comment by stuartmorgan** [comment]
> Flutter (Channel unknown, v1.9.1+hotfix.2, on Linux, locale en_US.UTF-8)

The problem is that you're using a 1.9.1, branch but not on a known channel (which ends up being treated as master). `precache --linux` is only supported on master. The artifact names changes, and versions branched before the change are looking for the wrong artifacts.

You should either switch to stable channel, in which case this would have correctly skipped the Linux artifact, or use an actual master (non-branch) version.

**Comment by jaggernod** [comment]
Thanks, @stuartmorgan for the answer. 
I'm using this method on my CI Docker image and it did work till v1.8.4.

I do want to be on a constant tag (not stable nor beta as I lose the build reproducibility). When I try o change to a tag with `flutter version v1.9.1+hotfix.2` and then run my Linux build I get the same type of error as described above.

```
flutter run -d Linux
Downloading linux-x64 tools...                                          
Download failed.
URL: https://storage.googleapis.com/flutter_infra/flutter/b863200c37df4ed378042de11c4e9ff34e4e58c9/linux-x64/linux-x64-flutter.zip
Error: 404 Not Found
```

**Comment by stuartmorgan** [comment]
> it did work till v1.8.4.

1.8.4 isn't a branch, so it's identical to using an old version of master (if you're not specifically on one of the special named branches: `stable`, `beta`, `dev`). Desktop should therefore work as well at that hash worked when it was the latest master.

> I do want to be on a constant tag (not stable nor beta as I lose the build reproducibility).

If you want to use a pinned version for desktop builds, you should pick a tag or hash that's not on a branch.

> When I try o change to a tag with `flutter version v1.9.1+hotfix.2` and then run my Linux build I get the same type of error as described above.

Yes, that's because that tag is on a branch (any `hotfix` version is), as I said in my earlier reply. A branch does not correspond to an earlier state of master, and desktop is supported only on master. (In this case, the problem is due to artifact builder recipes not being branched, which is not ideal, but not something we are going to work around to make desktop work on a branch where it's explicitly not supported.)

**Comment by jaggernod** [comment]
Just to sum up (for the people who might have a similar issue).

In order to have the same flutter version for both **mobile** and **desktop** (as in my case when I run Driver Test on Linux) one cannot use hotfix branches. 
The current solution is to use any commit that is in the **master** branch. 

In the provided case I need to use for example *v1.9.2* instead of *v1.9.1+hotfix.2*. Obviously it won't contain the cherry-picked fixes that are in *v1.9.1+hotfix.2*. But I'm fine with it.

@stuartmorgan maybe it would be nice if the error message for the `precache` failing on the non-master branch gave some more explanation. No way I would figure out what was happening without @stuartmorgan help.

**Comment by stuartmorgan** [comment]
> [...] one cannot use hotfix branches.

As a slight correction, the correct statement is that using desktop on anything other than master, including hotfix branches, is unsupported. It might work (in fact, it will probably work most of the time; this is an unusual edge case), but there are no guarantees.

> it would be nice if the error message for the `precache` failing on the non-master branch gave some more explanation.

The only reason you had this error in the first place is that you explicitly put your tree in the state of, on one hand, being on a branch that's only intended for `stable`/`beta` (where desktop isn't supported), while on the other hand not actually being on the `stable`/`beta` branch (which the tool checks for to disable desktop support, which would have avoided this error).

In theory, yes, it would be nice if the tool had a specific check that would have prevented this. In practice, creating even more convoluted logic specifically to handle this very rare case, and that *only* for people who are already working around existing protections in the tool (and who are not following the recommendation that people experimenting with the preliminary desktop support stay on the latest master in general), would not be a good investment of time.
